File: ios/Internal/Security/AccessControlResolver.swift
import Foundation
import Security

enum AccessControlResolverError: Error {
  case unableToCreateAccessControl
}

enum AccessPolicy: String {
  case secureEnclaveBiometry = "secureEnclaveBiometry"
  case biometryCurrentSet = "biometryCurrentSet"
  case biometryAny = "biometryAny"
  case devicePasscode = "devicePasscode"
  case none = "none"
}

enum SecurityTier: String {
  case secureEnclave = "secureEnclave"
  case strongBox = "strongBox"
  case biometry = "biometry"
  case deviceCredential = "deviceCredential"
  case software = "software"
}

struct AccessControlContext {
  let policy: AccessPolicy
  let securityLevel: SecurityTier
  let accessible: CFString
  let accessControlRef: SecAccessControl?
}

final class AccessControlResolver {
  private let resolveAvailability: () -> (secureEnclave: Bool, strongBox: Bool, biometry: Bool, deviceCredential: Bool)
  private let defaultOrder: [AccessPolicy] = [
    .secureEnclaveBiometry,
    .biometryCurrentSet,
    .biometryAny,
    .devicePasscode,
    .none
  ]

  init(resolveAvailability: @escaping () -> (secureEnclave: Bool, strongBox: Bool, biometry: Bool, deviceCredential: Bool)) {
    self.resolveAvailability = resolveAvailability
  }

  func resolve(preferred: AccessPolicy?) throws -> AccessControlContext {
    let availability = resolveAvailability()
    let ordered = orderedPreferences(preferred: preferred)

    for candidate in ordered {
      if let context = try attemptResolve(candidate, availability: availability) {
        return context
      }
    }

    return AccessControlContext(
      policy: .none,
      securityLevel: .software,
      accessible: kSecAttrAccessibleAfterFirstUnlock,
      accessControlRef: nil
    )
  }

  private func orderedPreferences(preferred: AccessPolicy?) -> [AccessPolicy] {
    guard let preferred else {
      return defaultOrder
    }
    if preferred == .none {
      return [.none]
    }
    return [preferred] + defaultOrder.filter { $0 != preferred }
  }

  private func attemptResolve(
    _ candidate: AccessPolicy,
  availability: (secureEnclave: Bool, strongBox: Bool, biometry: Bool, deviceCredential: Bool)
  ) throws -> AccessControlContext? {
    switch candidate {
    case .secureEnclaveBiometry:
      guard availability.secureEnclave, availability.deviceCredential else { return nil }
      let accessControl = try makeAccessControl(flags: [.biometryCurrentSet])
      return AccessControlContext(
        policy: .secureEnclaveBiometry,
        securityLevel: .secureEnclave,
        accessible: kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly,
        accessControlRef: accessControl
      )
    case .biometryCurrentSet:
      guard availability.biometry, availability.deviceCredential else { return nil }
      let accessControl = try makeAccessControl(flags: [.biometryCurrentSet])
      let security: SecurityTier = availability.secureEnclave ? .secureEnclave : .biometry
      return AccessControlContext(
        policy: .biometryCurrentSet,
        securityLevel: security,
        accessible: kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly,
        accessControlRef: accessControl
      )
    case .biometryAny:
      guard availability.biometry, availability.deviceCredential else { return nil }
      let accessControl = try makeAccessControl(flags: [.biometryAny])
      let security: SecurityTier = availability.secureEnclave ? .secureEnclave : .biometry
      return AccessControlContext(
        policy: .biometryAny,
        securityLevel: security,
        accessible: kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly,
        accessControlRef: accessControl
      )
    case .devicePasscode:
      guard availability.deviceCredential else { return nil }
      let accessControl = try makeAccessControl(flags: [.devicePasscode])
      return AccessControlContext(
        policy: .devicePasscode,
        securityLevel: .deviceCredential,
        accessible: kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly,
        accessControlRef: accessControl
      )
    case .none:
      return AccessControlContext(
        policy: .none,
        securityLevel: .software,
        accessible: kSecAttrAccessibleAfterFirstUnlock,
        accessControlRef: nil
      )
    }
  }

  private func makeAccessControl(flags: SecAccessControlCreateFlags) throws -> SecAccessControl {
    var error: Unmanaged<CFError>?
    guard let control = SecAccessControlCreateWithFlags(
      kCFAllocatorDefault,
      kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly,
      flags,
      &error
    ) else {
      if let existing = error?.takeRetainedValue() {
        throw existing
      }
      throw AccessControlResolverError.unableToCreateAccessControl
    }
    return control
  }
}

File: android/src/main/java/com/sensitiveinfo/internal/crypto/AccessControlResolver.kt
package com.sensitiveinfo.internal.crypto

import androidx.biometric.BiometricManager.Authenticators
import com.margelo.nitro.sensitiveinfo.AccessControl
import com.margelo.nitro.sensitiveinfo.SecurityLevel

/**
 * Determines which Android security primitives should back a requested access control.
 *
 * The resolver walks through a preference list, discarding options that are unavailable on the
 * current device (e.g. no StrongBox, weak biometrics only). The resulting `AccessResolution`
 * instructs the `CryptoManager` how to create or reopen the matching keystore key.
 */
internal class AccessControlResolver(
  private val availabilityResolver: SecurityAvailabilityResolver
) {
  private val defaultOrder = listOf(
    AccessControl.SECUREENCLAVEBIOMETRY,
    AccessControl.BIOMETRYCURRENTSET,
    AccessControl.BIOMETRYANY,
    AccessControl.DEVICEPASSCODE,
    AccessControl.NONE
  )

  /** Chooses the best available policy given the caller preference and hardware capabilities. */
  fun resolve(preferred: AccessControl?, strongBiometricsOnly: Boolean): AccessResolution {
    val availability = availabilityResolver.resolve()
    val ordered = orderPreferences(preferred)

    for (candidate in ordered) {
      val resolution = tryResolve(candidate, availability, strongBiometricsOnly)
      if (resolution != null) {
        return resolution
      }
    }

    return AccessResolution(
      accessControl = AccessControl.NONE,
      securityLevel = SecurityLevel.SOFTWARE,
      requiresAuthentication = false,
      allowedAuthenticators = 0,
      useStrongBox = false,
      invalidateOnEnrollment = false
    )
  }

  private fun orderPreferences(preferred: AccessControl?): List<AccessControl> {
    if (preferred == null) {
      return defaultOrder
    }
    if (preferred == AccessControl.NONE) {
      return listOf(AccessControl.NONE)
    }
    val remaining = defaultOrder.filter { it != preferred }
    return listOf(preferred) + remaining
  }

  private fun tryResolve(
    accessControl: AccessControl,
    availability: SecurityAvailabilitySnapshot,
    strongBiometricsOnly: Boolean
  ): AccessResolution? {
    return when (accessControl) {
      AccessControl.SECUREENCLAVEBIOMETRY -> {
        if (!availability.biometry || !availability.strongBox) return null
        AccessResolution(
          accessControl = AccessControl.SECUREENCLAVEBIOMETRY,
          securityLevel = SecurityLevel.STRONGBOX,
          requiresAuthentication = true,
          allowedAuthenticators = Authenticators.BIOMETRIC_STRONG,
          useStrongBox = true,
          invalidateOnEnrollment = true
        )
      }
      AccessControl.BIOMETRYCURRENTSET -> {
        if (!availability.biometry) return null
        AccessResolution(
          accessControl = AccessControl.BIOMETRYCURRENTSET,
          securityLevel = SecurityLevel.BIOMETRY,
          requiresAuthentication = true,
          allowedAuthenticators = if (strongBiometricsOnly) {
            Authenticators.BIOMETRIC_STRONG
          } else {
            Authenticators.BIOMETRIC_STRONG or Authenticators.BIOMETRIC_WEAK
          },
          useStrongBox = false,
          invalidateOnEnrollment = true
        )
      }
      AccessControl.BIOMETRYANY -> {
        if (!availability.biometry) return null
        AccessResolution(
          accessControl = AccessControl.BIOMETRYANY,
          securityLevel = SecurityLevel.BIOMETRY,
          requiresAuthentication = true,
          allowedAuthenticators = if (strongBiometricsOnly) {
            Authenticators.BIOMETRIC_STRONG
          } else {
            Authenticators.BIOMETRIC_STRONG or Authenticators.BIOMETRIC_WEAK
          },
          useStrongBox = false,
          invalidateOnEnrollment = false
        )
      }
      AccessControl.DEVICEPASSCODE -> {
        if (!availability.deviceCredential) return null
        AccessResolution(
          accessControl = AccessControl.DEVICEPASSCODE,
          securityLevel = SecurityLevel.DEVICECREDENTIAL,
          requiresAuthentication = true,
          allowedAuthenticators = Authenticators.DEVICE_CREDENTIAL,
          useStrongBox = false,
          invalidateOnEnrollment = false
        )
      }
      AccessControl.NONE -> AccessResolution(
        accessControl = AccessControl.NONE,
        securityLevel = SecurityLevel.SOFTWARE,
        requiresAuthentication = false,
        allowedAuthenticators = 0,
        useStrongBox = false,
        invalidateOnEnrollment = false
      )
    }
  }
}
